src/read/pipelining.rs

```diff
@@ -0,0 +1,1209 @@
+//! Pipelined extraction into a filesystem directory.
+
+pub mod path_splitting {
+    use displaydoc::Display;
+    use thiserror::Error;
+
+    use std::collections::BTreeMap;
+
+    use crate::spec::is_dir;
+
+    /// Errors encountered during path splitting.
+    #[derive(Debug, Display, Error)]
+    pub enum PathSplitError {
+        /// entry path format error: {0:?}
+        PathFormat(String),
+        /// file and directory paths overlapped: {0:?}
+        FileDirOverlap(String),
+    }
+
+    fn split_by_separator(entry_path: &str) -> Result<impl Iterator<Item = &str>, PathSplitError> {
+        if entry_path.contains('\\') {
+            if entry_path.contains('/') {
+                return Err(PathSplitError::PathFormat(format!(
+                    "path {:?} contained both '\\' and '/' separators",
+                    entry_path
+                )));
+            }
+            Ok(entry_path.split('\\'))
+        } else {
+            Ok(entry_path.split('/'))
+        }
+    }
+
+    /* TODO: consider using crate::unstable::path_to_string() for this--it involves new
+     * allocations, but that really shouldn't matter for our purposes. I like the idea of using our
+     * own logic here, since parallel/pipelined extraction is really a different use case than the
+     * rest of the zip crate, but it's definitely worth considering. */
+    pub(crate) fn normalize_parent_dirs<'a>(
+        entry_path: &'a str,
+    ) -> Result<(Vec<&'a str>, bool), PathSplitError> {
+        if entry_path.starts_with('/') || entry_path.starts_with('\\') {
+            return Err(PathSplitError::PathFormat(format!(
+                "path {:?} began with '/' or '\\' and is absolute",
+                entry_path
+            )));
+        }
+        let is_dir = is_dir(entry_path);
+
+        let mut ret: Vec<&'a str> = Vec::new();
+        for component in split_by_separator(entry_path)? {
+            match component {
+                /* Skip over repeated separators "//". We check separately for ending '/' with the
+                 * `is_dir` variable. */
+                "" => (),
+                /* Skip over redundant "." separators. */
+                "." => (),
+                /* If ".." is present, pop off the last element or return an error. */
+                ".." => {
+                    if ret.pop().is_none() {
```

Comment:
This is valid if the destination is the root directory, because `/..` is the same as `/`.

Reply:
Ah ok, so this is literally handling the case where instead of a path relative to the extraction dir, the entry is supposed to expand into an absolute path? From the current `.extract()` code where we call `.enclosed_name()`, it looks like we currently don't support absolute extraction paths? I think my intent here was to mimic that logic, but are you saying instead that we should convert absolute entry names beginning with `/` as relative to the extraction dir?

This code is purely processing entry names right now (not yet conjoined to the extraction dir), so I was under the impression they should all be relative paths, and that (like `.enclosed_name()`), if they use too many `..`s, that we should error out. I'm not sure how to square that with "if the destination is the root directory". This is probably pretty simple but I would appreciate further clarification here (thanks!).

Reply:
Well, the POSIX standard states that `/..` and `/` are the same directory. Unless the ZIP APPNOTE says otherwise, I want to follow that standard.